File: backend/src/main/java/cn/etarch/mao/session/ws/StreamingWsRegistry.java
package cn.etarch.mao.session.ws;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.annotation.PreDestroy;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.TextMessage;
import org.springframework.web.socket.WebSocketSession;

import java.io.IOException;
import java.util.Set;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.stream.Collectors;

@Slf4j
@Component
public class StreamingWsRegistry {

    private enum SendTarget {
        ALL,
        LOCAL_ONLY
    }

    private record OutboundItem(Long userId, WsEvent event, String rawJson, SendTarget target,
                                CompletableFuture<WsDeliveryResult> resultFuture) {}

    public record WsDeliveryResult(int targetCount, int successCount, int failureCount) {
        public boolean delivered() {
            return successCount > 0;
        }
    }

    private static OutboundItem eventItem(Long userId, WsEvent event, SendTarget target) {
        return new OutboundItem(userId, event, null, target, null);
    }

    private static OutboundItem rawItem(Long userId, String rawJson, SendTarget target) {
        return new OutboundItem(userId, null, rawJson, target, null);
    }

    private final ObjectMapper objectMapper = new ObjectMapper();
    private final BlockingQueue<OutboundItem> outboundQueue;
    private final AtomicBoolean running = new AtomicBoolean(true);
    private final Thread senderThread;

    /** userId → set of active WebSocket sessions (multi-device) */
    private final ConcurrentHashMap<Long, Set<WebSocketSession>> userSessions = new ConcurrentHashMap<>();

    /** WebSocket session → userId (reverse lookup) */
    private final ConcurrentHashMap<String, Long> sessionToUser = new ConcurrentHashMap<>();

    /** WebSocket session → client type (electron/browser) */
    private final ConcurrentHashMap<String, String> sessionToClientType = new ConcurrentHashMap<>();

    /** userId → set of subscribed sessionIds */
    private final ConcurrentHashMap<Long, Set<Long>> userSubscriptions = new ConcurrentHashMap<>();

    public StreamingWsRegistry(
            @Value("${app.ws.outbound-queue-capacity:10000}") int outboundQueueCapacity) {
        this.outboundQueue = new LinkedBlockingQueue<>(outboundQueueCapacity);
        this.senderThread = new Thread(this::drainOutboundLoop, "ws-outbound");
        this.senderThread.setDaemon(true);
        this.senderThread.start();
    }

    @PreDestroy
    void shutdown() {
        running.set(false);
        senderThread.interrupt();
    }

    public int getOutboundQueueSize() {
        return outboundQueue.size();
    }

    public void register(WebSocketSession session, Long userId, String clientType) {
        String sessionId = session.getId();
        sessionToUser.put(sessionId, userId);
        sessionToClientType.put(sessionId, normalizeClientType(clientType));
        userSessions.computeIfAbsent(userId, k -> ConcurrentHashMap.newKeySet()).add(session);
        log.info("WS stream registered: userId={}, wsSessionId={}, clientType={}",
                userId, sessionId, sessionToClientType.get(sessionId));
    }

    public void unregister(WebSocketSession session) {
        String sessionId = session.getId();
        Long userId = sessionToUser.remove(sessionId);
        sessionToClientType.remove(sessionId);
        if (userId != null) {
            Set<WebSocketSession> sessions = userSessions.get(userId);
            if (sessions != null) {
                sessions.remove(session);
                if (sessions.isEmpty()) {
                    userSessions.remove(userId);
                    userSubscriptions.remove(userId);
                }
            }
            log.info("WS stream unregistered: userId={}, wsSessionId={}", userId, sessionId);
        }
    }

    public void subscribe(Long userId, Long sessionId) {
        userSubscriptions.computeIfAbsent(userId, k -> ConcurrentHashMap.newKeySet()).add(sessionId);
    }

    public void unsubscribe(Long userId, Long sessionId) {
        Set<Long> subs = userSubscriptions.get(userId);
        if (subs != null) {
            subs.remove(sessionId);
        }
    }

    public boolean isSubscribed(Long userId, Long sessionId) {
        Set<Long> subs = userSubscriptions.get(userId);
        return subs != null && subs.contains(sessionId);
    }

    /**
     * Enqueue an event for async delivery — does not block the caller on network I/O.
     */
    public void send(Long userId, WsEvent event) {
        enqueue(userId, event, SendTarget.ALL);
    }

    public CompletableFuture<WsDeliveryResult> sendWithResult(Long userId, WsEvent event) {
        CompletableFuture<WsDeliveryResult> future = new CompletableFuture<>();
        if (userId == null || event == null) {
            future.complete(new WsDeliveryResult(0, 0, 0));
            return future;
        }
        if (!outboundQueue.offer(new OutboundItem(userId, event, null, SendTarget.ALL, future))) {
            log.warn("WS outbound queue full, dropping tracked event type={} for userId={}",
                    event.getType(), userId);
            future.complete(new WsDeliveryResult(0, 0, 0));
        }
        return future;
    }

    /**
     * Enqueue an event for Electron desktop connections only.
     */
    public void sendToLocalClients(Long userId, WsEvent event) {
        enqueue(userId, event, SendTarget.LOCAL_ONLY);
    }

    /**
     * Enqueue a raw JSON message for async delivery.
     */
    public void sendRaw(Long userId, String json) {
        enqueueRaw(userId, json, SendTarget.ALL);
    }

    private void enqueue(Long userId, WsEvent event, SendTarget target) {
        if (userId == null || event == null) {
            return;
        }
        if (!outboundQueue.offer(eventItem(userId, event, target))) {
            log.warn("WS outbound queue full (capacity reached), dropping event type={} for userId={}",
                    event.getType(), userId);
        }
    }

    private void enqueueRaw(Long userId, String json, SendTarget target) {
        if (userId == null || json == null) {
            return;
        }
        if (!outboundQueue.offer(rawItem(userId, json, target))) {
            log.warn("WS outbound queue full, dropping raw message for userId={}", userId);
        }
    }

    private void drainOutboundLoop() {
        while (running.get()) {
            try {
                OutboundItem item = outboundQueue.poll(1, TimeUnit.SECONDS);
                if (item != null) {
                    deliver(item);
                }
            } catch (InterruptedException e) {
                if (!running.get()) {
                    Thread.currentThread().interrupt();
                    break;
                }
            } catch (Exception e) {
                log.warn("WS outbound delivery error: {}", e.getMessage());
            }
        }
    }

    private void deliver(OutboundItem item) {
        Set<WebSocketSession> sessions = userSessions.get(item.userId());
        if (sessions == null || sessions.isEmpty()) {
            completeResult(item, 0, 0, 0);
            return;
        }

        Set<WebSocketSession> targets = switch (item.target()) {
            case ALL -> sessions;
            case LOCAL_ONLY -> sessions.stream()
                    .filter(session -> "electron".equals(sessionToClientType.get(session.getId())))
                    .collect(Collectors.toSet());
        };
        if (targets.isEmpty()) {
            completeResult(item, 0, 0, 0);
            return;
        }

        String json;
        if (item.rawJson() != null) {
            json = item.rawJson();
        } else {
            try {
                json = objectMapper.writeValueAsString(item.event());
            } catch (Exception e) {
                log.error("Failed to serialize WsEvent", e);
                completeResult(item, targets.size(), 0, targets.size());
                return;
            }
        }

        TextMessage message = new TextMessage(json);
        int targetCount = 0;
        int successCount = 0;
        int failureCount = 0;
        for (WebSocketSession session : targets) {
            if (session.isOpen()) {
                targetCount++;
                try {
                    synchronized (session) {
                        session.sendMessage(message);
                    }
                    successCount++;
                } catch (IOException e) {
                    failureCount++;
                    log.warn("Failed to send WS message to userId={}, wsSessionId={}: {}",
                            item.userId(), session.getId(), e.getMessage());
                }
            }
        }
        completeResult(item, targetCount, successCount, failureCount);
    }

    private void completeResult(OutboundItem item, int targets, int successes, int failures) {
        if (item.resultFuture() != null && !item.resultFuture().isDone()) {
            item.resultFuture().complete(new WsDeliveryResult(targets, successes, failures));
        }
    }

    public boolean hasConnection(Long userId) {
        Set<WebSocketSession> sessions = userSessions.get(userId);
        return sessions != null && sessions.stream().anyMatch(WebSocketSession::isOpen);
    }

    public boolean hasLocalClientConnection(Long userId) {
        Set<WebSocketSession> sessions = userSessions.get(userId);
        return sessions != null && sessions.stream()
                .anyMatch(session -> session.isOpen()
                        && "electron".equals(sessionToClientType.get(session.getId())));
    }

    /**
     * Get all subscribed session IDs for a user.
     */
    public Set<Long> getSubscribedSessionIds(Long userId) {
        Set<Long> subs = userSubscriptions.get(userId);
        return subs != null ? Set.copyOf(subs) : Set.of();
    }

    public Long getUserId(WebSocketSession session) {
        return sessionToUser.get(session.getId());
    }

    private String normalizeClientType(String clientType) {
        return "electron".equalsIgnoreCase(clientType) ? "electron" : "browser";
    }
}

File: backend-ts/src/session/ws/streaming-ws-registry.ts
import type { WsEvent } from './ws-event.js';

export interface WsSocket {
  id: string;
  readyState: number;
  send(data: string): void;
  close(code?: number, reason?: string): void;
}

export const WS_OPEN = 1;

/** 能作为 LOCAL 模式本机执行端的 WS client 类型。 */
export const LOCAL_CAPABLE_CLIENTS = new Set(['electron', 'cli']);

export interface WsDeliveryResult {
  targetCount: number;
  successCount: number;
  failureCount: number;
}

export function delivered(result: WsDeliveryResult): boolean {
  return result.successCount > 0;
}

export interface WsAuthMetadata {
  userId: number;
  authSource?: string;
  expiresAt: number;
}

interface ConnectionAuth {
  metadata: WsAuthMetadata;
  timer: ReturnType<typeof setTimeout> | null;
  refreshes: Map<string, number>;
}

type SendTarget = 'ALL' | 'LOCAL_ONLY';

interface OutboundItem {
  userId: number;
  event: WsEvent | null;
  rawJson: string | null;
  target: SendTarget;
  resultFuture: { resolve: (r: WsDeliveryResult) => void } | null;
}

export class StreamingWsRegistry {
  private readonly outboundQueue: OutboundItem[] = [];
  private readonly capacity: number;
  private running = true;
  private drainTimer: ReturnType<typeof setTimeout> | null = null;

  private readonly connectionAuth = new Map<string, ConnectionAuth>();
  private readonly closedConnections = new WeakSet<WsSocket>();
  private readonly userSessions = new Map<number, Set<WsSocket>>();
  private readonly sessionToUser = new Map<string, number>();
  private readonly sessionToClientType = new Map<string, string>();
  private readonly userSubscriptions = new Map<number, Set<number>>();
  private readonly activeToolCalls = new Map<number, Map<string, Record<string, unknown>>>();

  constructor(outboundQueueCapacity = 10000) {
    this.capacity = outboundQueueCapacity;
    this.scheduleDrain();
  }

  shutdown(): void {
    this.running = false;
    for (const auth of this.connectionAuth.values()) {
      if (auth.timer) clearTimeout(auth.timer);
    }
    this.connectionAuth.clear();
    if (this.drainTimer) {
      clearTimeout(this.drainTimer);
      this.drainTimer = null;
    }
  }

  getOutboundQueueSize(): number {
    return this.outboundQueue.length;
  }

  register(session: WsSocket, userId: number, clientType: string | null | undefined, metadata?: WsAuthMetadata): void {
    if (this.closedConnections.has(session)) return;
    const previous = this.connectionAuth.get(session.id);
    if (previous?.timer) clearTimeout(previous.timer);
    this.connectionAuth.delete(session.id);
    if (metadata) {
      const auth: ConnectionAuth = { metadata: { ...metadata }, timer: null, refreshes: new Map() };
      this.connectionAuth.set(session.id, auth);
      this.scheduleAuthExpiry(session, auth);
    }
    this.sessionToUser.set(session.id, userId);
    this.sessionToClientType.set(session.id, this.normalizeClientType(clientType));
    let set = this.userSessions.get(userId);
    if (!set) {
      set = new Set();
      this.userSessions.set(userId, set);
    }
    set.add(session);
    console.info(`WS stream registered: userId=${userId}, wsSessionId=${session.id}, clientType=${this.sessionToClientType.get(session.id)}`);
  }

  unregister(session: WsSocket): void {
    this.closedConnections.add(session);
    const auth = this.connectionAuth.get(session.id);
    if (auth?.timer) clearTimeout(auth.timer);
    this.connectionAuth.delete(session.id);
    const userId = this.sessionToUser.get(session.id);
    this.sessionToUser.delete(session.id);
    this.sessionToClientType.delete(session.id);
    if (userId != null) {
      const sessions = this.userSessions.get(userId);
      if (sessions) {
        sessions.delete(session);
        if (sessions.size === 0) {
          this.userSessions.delete(userId);
          this.userSubscriptions.delete(userId);
        }
      }
    }
  }

  /** 收发入口均检查墙上时间，不依赖到期回调获得调度。 */
  isConnectionAuthorized(session: WsSocket): boolean {
    if (this.closedConnections.has(session)) return false;
    const auth = this.connectionAuth.get(session.id);
    if (auth?.metadata.authSource === 'company_sso' && auth.metadata.expiresAt <= Date.now()) {
      this.closeConnection(session, 'Authentication expired');
      return false;
    }
    return true;
  }

  closeConnection(session: WsSocket, reason: string): void {
    if (this.closedConnections.has(session)) return;
    this.closedConnections.add(session);
    const auth = this.connectionAuth.get(session.id);
    if (auth?.timer) clearTimeout(auth.timer);
    this.connectionAuth.delete(session.id);
    session.close(1003, reason);
  }

  getRefreshResult(session: WsSocket, requestId: string): number | undefined {
    return this.connectionAuth.get(session.id)?.refreshes.get(requestId);
  }

  /** 调用方提供已验证元数据；同步完成校验、替换及计时器更新。 */
  refreshAuthentication(session: WsSocket, requestId: string, metadata: WsAuthMetadata): boolean {
    if (!this.isConnectionAuthorized(session)) return false;
    const auth = this.connectionAuth.get(session.id);
    if (!auth || metadata.userId !== auth.metadata.userId
      || metadata.authSource !== auth.metadata.authSource || metadata.expiresAt <= Date.now()) return false;
    if (auth.refreshes.has(requestId)) return true;
    auth.metadata = { ...metadata };
    auth.refreshes.set(requestId, metadata.expiresAt);
    this.scheduleAuthExpiry(session, auth);
    return true;
  }

  /** 认证确认只属于当前连接，不能广播到同一用户的其他客户端。 */
  sendToConnection(session: WsSocket, frame: unknown): void {
    if (this.isConnectionAuthorized(session) && session.readyState === WS_OPEN) {
      session.send(JSON.stringify(frame));
    }
  }

  private scheduleAuthExpiry(session: WsSocket, auth: ConnectionAuth): void {
    if (auth.timer) clearTimeout(auth.timer);
    auth.timer = null;
    if (auth.metadata.authSource !== 'company_sso') return;
    auth.timer = setTimeout(() => {
      auth.timer = null;
      if (this.isConnectionAuthorized(session)) this.scheduleAuthExpiry(session, auth);
    }, Math.max(0, Math.min(auth.metadata.expiresAt - Date.now(), 2_147_483_647)));
    auth.timer.unref?.();
  }

  subscribe(userId: number, sessionId: number): void {
    let set = this.userSubscriptions.get(userId);
    if (!set) {
      set = new Set();
      this.userSubscriptions.set(userId, set);
    }
    set.add(sessionId);
  }

  unsubscribe(userId: number, sessionId: number): void {
    this.userSubscriptions.get(userId)?.delete(sessionId);
  }

  isSubscribed(userId: number, sessionId: number): boolean {
    return this.userSubscriptions.get(userId)?.has(sessionId) ?? false;
  }

  trackActiveToolCall(sessionId: number, executionId: string, toolCallId: string, toolName: string, args: string): void {
    if (!toolCallId) return;
    let calls = this.activeToolCalls.get(sessionId);
    if (!calls) {
      calls = new Map();
      this.activeToolCalls.set(sessionId, calls);
    }
    calls.set(toolCallId, {
      tool_call_id: toolCallId,
      tool_name: toolName,
      arguments: args,
      executionId,
    });
  }

  updateActiveToolCallArguments(sessionId: number, toolCallId: string, args: string): void {
    const call = this.activeToolCalls.get(sessionId)?.get(toolCallId);
    if (call) call.arguments = args;
  }

  completeActiveToolCall(sessionId: number, toolCallId: string): void {
    const calls = this.activeToolCalls.get(sessionId);
    if (!calls) return;
    calls.delete(toolCallId);
    if (calls.size === 0) this.activeToolCalls.delete(sessionId);
  }

  getActiveToolCalls(sessionId: number): Array<Record<string, unknown>> {
    return [...(this.activeToolCalls.get(sessionId)?.values() ?? [])].map((call) => ({ ...call }));
  }

  clearActiveToolCalls(sessionId: number): void {
    this.activeToolCalls.delete(sessionId);
  }

  send(userId: number, event: WsEvent): void {
    this.enqueue(userId, event, 'ALL');
  }

  sendWithResult(userId: number, event: WsEvent): Promise<WsDeliveryResult> {
    return new Promise((resolve) => {
      if (userId == null || event == null) {
        resolve({ targetCount: 0, successCount: 0, failureCount: 0 });
        return;
      }
      if (this.outboundQueue.length >= this.capacity) {
        console.warn(`WS outbound queue full, dropping tracked event type=${event.type} for userId=${userId}`);
        resolve({ targetCount: 0, successCount: 0, failureCount: 0 });
        return;
      }
      this.outboundQueue.push({ userId, event, rawJson: null, target: 'ALL', resultFuture: { resolve } });
      this.flushNow();
    });
  }

  sendToLocalClients(userId: number, event: WsEvent): void {
    this.enqueue(userId, event, 'LOCAL_ONLY');
  }

  sendRaw(userId: number, json: string): void {
    if (userId == null || json == null) return;
    if (this.outboundQueue.length >= this.capacity) {
      console.warn(`WS outbound queue full, dropping raw message for userId=${userId}`);
      return;
    }
    this.outboundQueue.push({ userId, event: null, rawJson: json, target: 'ALL', resultFuture: null });
  }

  hasConnection(userId: number): boolean {
    const sessions = this.userSessions.get(userId);
    return sessions != null && [...sessions].some((s) => s.readyState === WS_OPEN && this.isConnectionAuthorized(s));
  }

  hasLocalClientConnection(userId: number): boolean {
    const sessions = this.userSessions.get(userId);
    return sessions != null && [...sessions].some(
      (s) => s.readyState === WS_OPEN && this.isConnectionAuthorized(s) && LOCAL_CAPABLE_CLIENTS.has(this.sessionToClientType.get(s.id) ?? ''),
    );
  }

  getSubscribedSessionIds(userId: number): Set<number> {
    const subs = this.userSubscriptions.get(userId);
    return subs ? new Set(subs) : new Set();
  }

  getUserId(session: WsSocket): number | null {
    return this.sessionToUser.get(session.id) ?? null;
  }

  private enqueue(userId: number, event: WsEvent, target: SendTarget): void {
    if (userId == null || event == null) return;
    if (this.outboundQueue.length >= this.capacity) {
      console.warn(`WS outbound queue full (capacity reached), dropping event type=${event.type} for userId=${userId}`);
      return;
    }
    this.outboundQueue.push({ userId, event, rawJson: null, target, resultFuture: null });
    this.flushNow();
  }

  private scheduleDrain(): void {
    if (!this.running) return;
    this.drainTimer = setTimeout(() => {
      this.flushNow();
      this.scheduleDrain();
    }, 50);
  }

  private flushNow(): void {
    while (this.outboundQueue.length > 0) {
      const item = this.outboundQueue.shift()!;
      this.deliver(item);
    }
  }

  private deliver(item: OutboundItem): void {
    const sessions = this.userSessions.get(item.userId);
    if (!sessions || sessions.size === 0) {
      this.completeResult(item, 0, 0, 0);
      return;
    }
    const targets = item.target === 'ALL'
      ? [...sessions]
      : [...sessions].filter((s) => LOCAL_CAPABLE_CLIENTS.has(this.sessionToClientType.get(s.id) ?? ''));
    if (targets.length === 0) {
      this.completeResult(item, 0, 0, 0);
      return;
    }
    let json: string;
    if (item.rawJson != null) {
      json = item.rawJson;
    } else {
      try {
        json = JSON.stringify(item.event);
      } catch {
        this.completeResult(item, targets.length, 0, targets.length);
        return;
      }
    }
    let targetCount = 0;
    let successCount = 0;
    let failureCount = 0;
    for (const session of targets) {
      if (session.readyState === WS_OPEN && this.isConnectionAuthorized(session)) {
        targetCount++;
        try {
          session.send(json);
          successCount++;
        } catch {
          failureCount++;
        }
      }
    }
    this.completeResult(item, targetCount, successCount, failureCount);
  }

  private completeResult(item: OutboundItem, targetCount: number, successCount: number, failureCount: number): void {
    item.resultFuture?.resolve({ targetCount, successCount, failureCount });
  }

  private normalizeClientType(clientType: string | null | undefined): string {
    if (clientType?.toLowerCase() === 'electron') return 'electron';
    if (clientType?.toLowerCase() === 'android') return 'android';
    if (clientType?.toLowerCase() === 'cli') return 'cli';
    // 与 StreamingWsHandler.normalizeClient 口径一致：embed 连接不应被记成 browser
    if (clientType?.toLowerCase() === 'embed') return 'embed';
    return 'browser';
  }
}
